src/test/regress/sql/multi_truncate.sql

```diff
@@ -216,10 +227,102 @@ ROLLBACK;
 SELECT * FROM test_local_truncate;
 
 -- Undistribute table
-SELECT master_drop_all_shards('test_local_truncate', 'pubic', 'test_local_truncate');
+SELECT master_drop_all_shards('test_local_truncate', 'public', 'test_local_truncate');
 DELETE FROM pg_dist_partition WHERE logicalrelid = 'test_local_truncate'::regclass;
 
 -- Ensure local data is not truncated
 SELECT * FROM test_local_truncate;
 
+DROP TABLE test_local_truncate;
+
+-- Test truncate_local_data_after_distributing_table UDF
+CREATE TABLE referenced_table(id int UNIQUE, test_column int);
+INSERT INTO referenced_table SELECT x,x FROM generate_series(1,10000) x;
+
+CREATE TABLE referencing_table(id int, ref_id int REFERENCES referenced_table(id));
+INSERT INTO referencing_table SELECT * FROM referenced_table;
+
+-- The following will fail as the table is not distributed
+SELECT truncate_local_data_after_distributing_table('referenced_table');
+
+-- Test foreign keys from local tables to distributed tables
+-- We can not truncate local tables until all the local foreign keys are removed.
+SELECT create_distributed_table('referenced_table', 'id');
+
+-- The following will fail, as local referencing_table has fk references
+SELECT truncate_local_data_after_distributing_table('referenced_table');
+
+-- Test foreign keys between distributed tables
+SELECT create_distributed_table('referencing_table', 'ref_id');
+
+BEGIN;
+-- The following will no longer fail as the referencing table is now distributed
+SELECT truncate_local_data_after_distributing_table('referenced_table');
+SELECT * FROM table_sizes;
+ROLLBACK;
+
+-- observe that none of the tables are truncated
+SELECT * FROM table_sizes;
+
+-- test that if we truncate the referencing table, only said table is affected
+BEGIN;
+SELECT truncate_local_data_after_distributing_table('referencing_table');
+SELECT * FROM table_sizes;
+ROLLBACK;
+
+-- however if we truncate referenced table, both of the tables get truncated
+-- because we supply the CASCADE option
+-- test that if we truncate the referencing table, only said table is affected
+BEGIN;
+SELECT truncate_local_data_after_distributing_table('referenced_table');
+SELECT * FROM table_sizes;
+ROLLBACK;
+
+DROP TABLE referencing_table;
+DROP TABLE referenced_table;
+
+-- test truncating reference tables
+CREATE TABLE ref(id int UNIQUE, data int);
+INSERT INTO ref SELECT x,x FROM generate_series(1,10000) x;
+SELECT create_reference_table('ref');
+
+CREATE TABLE dist(id int, ref_id int REFERENCES ref(id));
+INSERT INTO dist SELECT * FROM ref;
+SELECT create_distributed_table('dist','id');
```

Comment:
Maybe add one more truncate_local_data_after_distributing_table('ref') test (possibly with rollback) before this line to test the caching issue.

Reply:
Added the test without a rollback, outside a transaction block.

The UDF returned an error, and we did not truncate anything as I expected.